src/dsa/rpo_stark/stark/air.rs

```diff
@@ -0,0 +1,223 @@
+
+        enforce_rpo_round(frame, result, periodic_values);
+    }
+
+    fn get_assertions(&self) -> Vec<Assertion<Self::BaseField>> {
+        // Assert that the public key is the correct one
+        let initial_step = 0;
+        let last_step = self.trace_length() - 1;
+        vec![
+            Assertion::single(0, initial_step, Self::BaseField::ZERO),
+            Assertion::single(1, initial_step, Self::BaseField::ZERO),
+            Assertion::single(2, initial_step, Self::BaseField::ZERO),
+            Assertion::single(3, initial_step, Self::BaseField::ZERO),
+            Assertion::single(8, initial_step, Self::BaseField::ZERO),
+            Assertion::single(9, initial_step, Self::BaseField::ZERO),
+            Assertion::single(10, initial_step, Self::BaseField::ZERO),
+            Assertion::single(11, initial_step, Self::BaseField::ZERO),
+            Assertion::single(4, last_step, self.pub_key[0]),
+            Assertion::single(5, last_step, self.pub_key[1]),
+            Assertion::single(6, last_step, self.pub_key[2]),
+            Assertion::single(7, last_step, self.pub_key[3]),
+        ]
+    }
+
+    fn get_periodic_column_values(&self) -> Vec<Vec<Self::BaseField>> {
+        get_round_constants()
+    }
+}
+
+pub struct PublicInputs {
+    pub(crate) pub_key: Word,
+    pub(crate) msg: Word,
+}
+
+impl PublicInputs {
+    pub fn new(pub_key: Word, msg: Word) -> Self {
+        Self { pub_key, msg }
+    }
+}
+
+impl ToElements<BaseElement> for PublicInputs {
+    fn to_elements(&self) -> Vec<BaseElement> {
+        let mut res = self.pub_key.to_vec();
+        res.extend_from_slice(self.msg.as_ref());
+        res
+    }
+}
+
+// HELPER EVALUATORS
+// ------------------------------------------------------------------------------------------------
+
+/// Enforces constraints for a single round of the Rescue Prime Optimized hash functions.
+pub fn enforce_rpo_round<E: FieldElement + From<BaseElement>>(
+    frame: &EvaluationFrame<E>,
+    result: &mut [E],
+    ark: &[E],
+) {
+    // compute the state that should result from applying the first 5 operations of the RPO round to
+    // the current hash state.
+    let mut step1 = [E::ZERO; STATE_WIDTH];
+    step1.copy_from_slice(frame.current());
+
+    apply_mds(&mut step1);
+    // add constants
+    for i in 0..STATE_WIDTH {
+        step1[i] += ark[i];
+    }
+    apply_sbox(&mut step1);
+    apply_mds(&mut step1);
+    // add constants
+    for i in 0..STATE_WIDTH {
+        step1[i] += ark[STATE_WIDTH + i];
+    }
+
+    // compute the state that should result from applying the inverse of the last operation of the
+    // RPO round to the next step of the computation.
+    let mut step2 = [E::ZERO; STATE_WIDTH];
+    step2.copy_from_slice(frame.next());
+    apply_sbox(&mut step2);
+
+    // make sure that the results are equal.
+    for i in 0..STATE_WIDTH {
+        result.agg_constraint(i, are_equal(step2[i], step1[i]));
```

Comment:
Do we actually need `agg_constraint` here? Wouldn't `result[i] = are_equal(step2[i], step1[i])` work just as well?

Reply:
No we don't, it is only used in one spot so we can just inline it instead